File: src-tauri/src/events.rs
//! Names of the Tauri event channels emitted by the Rust core.
//! Mirrored in `src/lib/events.ts`.

/// `InstallOutputEvent` — one line of stdout/stderr from a running install job.
pub const INSTALL_OUTPUT: &str = "install://output";
/// `InstallDoneEvent` — an install job finished (success, failure or cancel).
pub const INSTALL_DONE: &str = "install://done";
/// `DownloadProgressEvent` — bytes downloaded so far for a download job.
pub const DOWNLOAD_PROGRESS: &str = "download://progress";
/// `CheckResult` — an individual environment check completed (streamed while `run_env_checks` runs).
pub const CHECK_PROGRESS: &str = "checks://progress";
/// `InstallOutputEvent` — one line from the optional Codex Fast UI toolkit (ADR-0007).
pub const FAST_UI_OUTPUT: &str = "fastui://output";
/// `InstallDoneEvent` — a Codex Fast UI toolkit job finished.
pub const FAST_UI_DONE: &str = "fastui://done";

File: src/lib/events.ts
/**
 * Tauri event channel names emitted by the Rust core (mirror of `src-tauri/src/events.rs`),
 * plus typed `listen` helpers.
 */
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type {
  CheckResult,
  DownloadProgressEvent,
  InstallDoneEvent,
  InstallOutputEvent,
} from "./types";

export const EVENTS = {
  installOutput: "install://output",
  installDone: "install://done",
  downloadProgress: "download://progress",
  checkProgress: "checks://progress",
  fastUiOutput: "fastui://output",
  fastUiDone: "fastui://done",
} as const;

export const onInstallOutput = (cb: (e: InstallOutputEvent) => void): Promise<UnlistenFn> =>
  listen<InstallOutputEvent>(EVENTS.installOutput, (ev) => cb(ev.payload));

export const onInstallDone = (cb: (e: InstallDoneEvent) => void): Promise<UnlistenFn> =>
  listen<InstallDoneEvent>(EVENTS.installDone, (ev) => cb(ev.payload));

export const onDownloadProgress = (cb: (e: DownloadProgressEvent) => void): Promise<UnlistenFn> =>
  listen<DownloadProgressEvent>(EVENTS.downloadProgress, (ev) => cb(ev.payload));

export const onCheckProgress = (cb: (e: CheckResult) => void): Promise<UnlistenFn> =>
  listen<CheckResult>(EVENTS.checkProgress, (ev) => cb(ev.payload));

/** Codex Fast UI toolkit output — same payload shape as the install channels (ADR-0007). */
export const onFastUiOutput = (cb: (e: InstallOutputEvent) => void): Promise<UnlistenFn> =>
  listen<InstallOutputEvent>(EVENTS.fastUiOutput, (ev) => cb(ev.payload));

export const onFastUiDone = (cb: (e: InstallDoneEvent) => void): Promise<UnlistenFn> =>
  listen<InstallDoneEvent>(EVENTS.fastUiDone, (ev) => cb(ev.payload));
